src/ossl/aes.rs

```diff
@@ -1660,3 +1658,232 @@ impl Decryption for AesOperation {
+                }
+                let val: usize =
+                    unsafe { *(mech.pParameter as CK_MAC_GENERAL_PARAMS_PTR) }
+                        as usize;
+                if val > AES_BLOCK_SIZE {
+                    return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+                }
+                val
+            }
+            CKM_AES_CMAC => {
+                if mech.ulParameterLen != 0 {
+                    return err_rv!(CKR_ARGUMENTS_BAD);
+                }
+                AES_BLOCK_SIZE
+            }
+            _ => return err_rv!(CKR_MECHANISM_INVALID),
+        };
+        let mackey = object_to_raw_key(key)?;
+        let mut mac = match EvpMac::from_ptr(unsafe {
+            EVP_MAC_fetch(
+                get_libctx(),
+                name_as_char(OSSL_MAC_NAME_CMAC),
+                std::ptr::null(),
+            )
+        }) {
+            Ok(em) => em,
+            Err(_) => return err_rv!(CKR_DEVICE_ERROR),
+        };
+        let mut ctx = match EvpMacCtx::from_ptr(unsafe {
+            EVP_MAC_CTX_new(mac.as_mut_ptr())
+        }) {
+            Ok(emc) => emc,
+            Err(_) => return err_rv!(CKR_DEVICE_ERROR),
+        };
+        let params = OsslParam::new()
+            .add_const_c_string(
+                name_as_char(OSSL_MAC_PARAM_CIPHER),
+                match mackey.raw.len() {
+                    16 => name_as_char(CIPHER_NAME_AES128),
+                    24 => name_as_char(CIPHER_NAME_AES192),
+                    32 => name_as_char(CIPHER_NAME_AES256),
+                    _ => return err_rv!(CKR_KEY_INDIGESTIBLE),
+                },
+            )?
+            .finalize();
+
+        if unsafe {
+            EVP_MAC_init(
+                ctx.as_mut_ptr(),
+                mackey.raw.as_ptr(),
+                mackey.raw.len(),
+                params.as_ptr(),
+            )
+        } != 1
+        {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+        Ok(AesCmacOperation {
+            finalized: false,
+            in_use: false,
+            _key: mackey,
+            _mac: mac,
+            ctx: ctx,
+            maclen: maclen,
+        })
+    }
+
+    fn begin(&mut self) -> KResult<()> {
+        if self.in_use {
+            return err_rv!(CKR_OPERATION_NOT_INITIALIZED);
+        }
+        Ok(())
+    }
+
+    fn update(&mut self, data: &[u8]) -> KResult<()> {
+        if self.finalized {
+            return err_rv!(CKR_OPERATION_NOT_INITIALIZED);
+        }
+        self.in_use = true;
+
+        if unsafe {
+            EVP_MAC_update(self.ctx.as_mut_ptr(), data.as_ptr(), data.len())
+        } != 1
+        {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+
+        Ok(())
+    }
+
+    fn finalize(&mut self, output: &mut [u8]) -> KResult<()> {
+        if self.finalized {
+            return err_rv!(CKR_OPERATION_NOT_INITIALIZED);
+        }
+        /* It is valid to finalize without any update */
+        self.in_use = true;
```

Comment:
should this go to `in_use = false` to allow new `begin()`?

Reply:
(same in the aes_mac.rs)

Reply:
No, a new begin MUST NOT be allowed, begin() is used to "protect" one shot operations from being initiated after update() operations have already been started.

A new one shot operation require cancelling of finalizing the current operation and then initiating a new one via C_SignInit